File: project-2-coldstart/python/tasks/store.py
"""Data layer: JSON-file backed task store. Tasks have id, title, done, tags."""
import json
import os

DB_PATH = os.path.join(os.path.dirname(__file__), "..", "tasks.json")


def _load(db_path=DB_PATH):
    if not os.path.exists(db_path):
        return []
    with open(db_path, encoding="utf-8") as fh:
        return json.load(fh)


def _save(rows, db_path=DB_PATH):
    with open(db_path, "w", encoding="utf-8") as fh:
        json.dump(rows, fh, indent=2)


def add_task(title, tags=(), db_path=DB_PATH):
    rows = _load(db_path)
    task = {
        "id": max([r["id"] for r in rows], default=0) + 1,
        "title": title,
        "done": False,
        "tags": sorted(set(tags)),
    }
    rows.append(task)
    _save(rows, db_path)
    return task


def list_tasks(tag=None, db_path=DB_PATH):
    rows = _load(db_path)
    if tag is None:
        return rows
    return [r for r in rows if tag in r.get("tags", [])]


def get_task(task_id, db_path=DB_PATH):
    for row in _load(db_path):
        if row["id"] == task_id:
            return row
    raise KeyError(f"no task {task_id}")


def set_done(task_id, done=True, db_path=DB_PATH):
    rows = _load(db_path)
    for row in rows:
        if row["id"] == task_id:
            row["done"] = done
            _save(rows, db_path)
            return row
    raise KeyError(f"no task {task_id}")

File: project-2-coldstart/js/src/store.mjs
// Data layer: JSON-file backed task store.
import { readFileSync, writeFileSync, existsSync } from "node:fs";

export const DB_PATH = new URL("../tasks.json", import.meta.url).pathname;

export function load(dbPath = DB_PATH) {
  if (!existsSync(dbPath)) return [];
  return JSON.parse(readFileSync(dbPath, "utf-8"));
}

export function save(rows, dbPath = DB_PATH) {
  writeFileSync(dbPath, JSON.stringify(rows, null, 2));
}

export function addTask(title, tags = [], dbPath = DB_PATH) {
  const rows = load(dbPath);
  const task = {
    id: rows.reduce((m, r) => Math.max(m, r.id), 0) + 1,
    title,
    done: false,
    tags: [...new Set(tags)].sort(),
  };
  rows.push(task);
  save(rows, dbPath);
  return task;
}

export function listTasks(tag = null, dbPath = DB_PATH) {
  const rows = load(dbPath);
  if (tag === null) return rows;
  return rows.filter((r) => (r.tags ?? []).includes(tag));
}

export function getTask(id, dbPath = DB_PATH) {
  const found = load(dbPath).find((r) => r.id === id);
  if (!found) throw new Error(`no task ${id}`);
  return found;
}

export function setDone(id, done = true, dbPath = DB_PATH) {
  const rows = load(dbPath);
  const task = rows.find((r) => r.id === id);
  if (!task) throw new Error(`no task ${id}`);
  task.done = done;
  save(rows, dbPath);
  return task;
}
